collector/clickhouse.go

```diff
@@ -0,0 +1,280 @@
+func NewClickhouse(opts ClickhouseOpts) (*Clickhouse, error) {
+	ch := &Clickhouse{
+		log:                   opts.Log,
+		opts:                  opts,
+		currentTxBatch:        make([]common.TxSummaryEntry, 0, clickhouseBatchSize),
+		currentSourcelogBatch: make([]SourceLogEntry, 0, clickhouseBatchSize),
+	}
+	if ch.opts.DSN == "" {
+		return nil, ErrNoDSN
+	}
+
+	err := ch.connect()
+	if err != nil {
+		return nil, err
+	}
+	return ch, nil
+}
+
+// connect establishes a connection to the Clickhouse database using secure native protocol.
+func (ch *Clickhouse) connect() error {
+	ctx := context.Background()
+	var err error
+
+	// Parse the DSN to extract address and authentication details
+	options, err := clickhouse.ParseDSN(ch.opts.DSN)
+	if err != nil {
+		metrics.IncClickhouseError()
+		return fmt.Errorf("failed to parse Clickhouse DSN: %w", err)
+	}
+
+	var chTLS *tls.Config
+	if clickhouseEnableTLS {
+		chTLS = &tls.Config{
+			InsecureSkipVerify: true, //nolint:gosec
+		}
+	}
+
+	ch.conn, err = clickhouse.Open(&clickhouse.Options{
+		Addr: options.Addr,
+		Auth: options.Auth,
+		TLS:  chTLS,
+		Debugf: func(format string, v ...interface{}) {
+			ch.log.Infof("Clickhouse debug: "+format, v...)
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := ch.conn.Ping(ctx); err != nil {
+		metrics.IncClickhouseError()
+		ch.log.Errorw("Failed to connect to Clickhouse", "error", err)
+		return err
+	}
+
+	return ch.applyMigrations()
+}
+
+func (ch *Clickhouse) applyMigrations() error {
+	if clickhouseApplySQLPath == "" {
+		return nil
+	}
+
+	// Load files from disk
+	items, _ := os.ReadDir(clickhouseApplySQLPath)
+	for _, item := range items {
+		if item.IsDir() {
+			continue
+		}
+
+		filePath := clickhouseApplySQLPath + "/" + item.Name()
+		ch.log.Infow("Applying Clickhouse SQL file", "file", filePath)
+
+		sql, err := os.ReadFile(filePath)
+		if err != nil {
+			metrics.IncClickhouseError()
+			return fmt.Errorf("failed to read Clickhouse SQL file %s: %w", filePath, err)
+		}
+
+		if err := ch.conn.Exec(context.Background(), string(sql)); err != nil {
+			metrics.IncClickhouseError()
+			return fmt.Errorf("failed to execute Clickhouse SQL file %s: %w", filePath, err)
+		}
+	}
+
+	return nil
+}
+
+// AddTransaction adds a transaction to the Clickhouse batch. If the batch size exceeds the configured limit, it sends the batch to Clickhouse.
+func (ch *Clickhouse) AddTransaction(tx common.TxIn) error {
+	txSummary, _, err := common.ParseTx(0, tx.Tx)
+	if err != nil {
+		metrics.IncClickhouseError()
+		return fmt.Errorf("failed to parse transaction: %w", err)
+	}
+
+	// First, check if the current batch is full, in which case we need to send it to Clickhouse
+	// ch.currentTxBatchLock.Lock()
+	// defer ch.currentTxBatchLock.Unlock()
+	if len(ch.currentTxBatch) >= cap(ch.currentTxBatch) {
+		// Create a copy of the batche and save it to Clickhouse (with retries)
+		txs := make([]common.TxSummaryEntry, clickhouseBatchSize)
```

Comment:
I don't think it's possible for `len` to be greater than `cap`, maybe it's more obvious to compare with `clickhouseBatchSize`?

```suggestion
	if len(ch.currentTxBatch) >= clickhouseBatchSize {
		// Create a copy of the batche and save it to Clickhouse (with retries)
		txs := make([]common.TxSummaryEntry, len(ch.currentTxBatch))
```